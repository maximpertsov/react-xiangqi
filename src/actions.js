export {
  addMove,
  cancelMoves,
  confirmMoves,
  fetchInitialPlacement,
  fetchMoveInfo,
  makeMove,
  postMove,
  selectMove,
  toggleMovesFetched,
  toggleLoading,
} from 'scenes/Game/actions';

export {
  fetchGames,
  setAutoMoveOff,
  setAutoMoveRed,
  setAutoMoveBlack,
  setAutoMoveBoth,
  setGameSlug,
  setForm,
  setUsername,
  toggleShowGame,
  toggleCanMoveBothColors,
} from 'scenes/Home/actions';

export {
  clearAnimationOffset,
  clearSelectedSlot,
  setAnimationOffset,
  setSelectedSlot,
} from 'components/Board/actions';

export default {};
