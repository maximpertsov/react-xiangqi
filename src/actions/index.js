import { createActions } from 'redux-actions';
import { Color } from 'services/logic/constants';

export default createActions({
  BOARD: {
    ANIMATION_OFFSET: {
      SET: undefined,
      CLEAR: () => [0, 0],
    },
    SELECTED_SQUARE: {
      SET: undefined,
    },
  },
  GAME: {
    CAN_MOVE_BOTH_COLORS: {
      SET: undefined,
    },
    REQUESTED_TAKEBACK: {
      SET: undefined,
    },
    PLAYERS: {
      SET: undefined,
    },
    MOVES: {
      ADD: undefined,
      REMOVE: undefined,
      UPDATE: undefined,
      SET: undefined,
    },
    SELECTED_MOVE: {
      SET: undefined,
    },
    SHOW_CONFIRM_MOVE_MENU: {
      SET: undefined,
    },
    SLUG: {
      SET: undefined,
    },
    UPDATE_COUNT: {
      SET: undefined,
    },
  },
  HOME: {
    AUTO_MOVE: {
      SET: {
        OFF: () => [],
        RED: () => [Color.RED],
        BLACK: () => [Color.BLACK],
        BOTH: () => [Color.RED, Color.BLACK],
      },
    },
    GAMES: {
      SET: undefined,
    },
    LOGIN_FORM: {
      USERNAME: {
        SET: undefined,
      },
      PASSWORD: {
        SET: undefined,
      },
      ERROR: {
        SET: undefined,
      },
      LOADING: {
        SET: undefined,
      },
    },
    SHOW_GAME: {
      SET: undefined,
    },
    USERNAME: {
      SET: undefined,
    },
  },
});
