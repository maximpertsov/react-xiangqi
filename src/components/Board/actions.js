export const setSelectedSlot = ({ selectedSlot }) => ({
  type: 'set_selected_slot',
  selectedSlot,
});

export const clearSelectedSlot = () => setSelectedSlot({ selectedSlot: null });

export const setAnimationOffset = ({
  bottomPlayerIsRed,
  fromSlot,
  toSlot,
}) => ({
  type: 'set_animation_offset',
  bottomPlayerIsRed,
  fromSlot,
  toSlot,
});

export const clearAnimationOffset = () => ({
  type: 'clear_animation_offset',
});

export default {};
