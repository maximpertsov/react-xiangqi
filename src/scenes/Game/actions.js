import update from 'immutability-helper';
import XiangqiBoard, * as logic from 'services/logic';
import { Color } from 'services/logic/constants';

// TODO: find out why this id starts at 3
let nextMoveId = 0;

const newMove = (moveData) => (
  {
    ...moveData,
    id: ++nextMoveId,
  }
);

/*********************/
/*** Initial State ***/
/*********************/

// TODO: define in logic class
/* eslint-disable-next-line max-len */
const DEFAULT_FEN = 'rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR';

const getInitialMovesState = (fen = DEFAULT_FEN) => ([
  newMove({
    piece: undefined,
    fromPos: undefined,
    toPos: undefined,
    board: new XiangqiBoard({ fen }),
    pending: false,
  }),
]);

const initialPlayers = [
  { name: undefined, color: Color.RED },
  { name: undefined, color: Color.BLACK },
];

export const getInitialState = () => ({
  loading: true,
  moves: getInitialMovesState(),
  moveCount: 0,
  players: initialPlayers,
  selectedMoveId: undefined,
});

/***************/
/*** Actions ***/
/***************/

const incrementMoveCount = (state) => (
  { ...state, moveCount: state.moveCount + 1 }
);

const getMoveIndex = ({ moves }, moveId ) => {
  if (moveId === undefined) return 0;

  return moves.findIndex(({ id }) => id === moveId);
};

const setSelectedMoveIndex = (state, moveIndex) => {
  const { id: selectedMoveId } = state.moves[moveIndex];
  return { ...state, selectedMoveId };
};

export const setSelectedMove = (state, moveId) =>
  setSelectedMoveIndex(state, getMoveIndex(state, moveId));

export const setPreviousMove = (state) => {
  const moveIndex = getMoveIndex(state, state.selectedMoveId);
  return setSelectedMoveIndex(state, Math.max(moveIndex - 1, 0));
};

export const setNextMove = (state) => {
  const moveIndex = getMoveIndex(state, state.selectedMoveId);
  return setSelectedMoveIndex(state, Math.min(moveIndex + 1, state.moveCount));
};

const selectLastMove = (state) =>
  setSelectedMoveIndex(state, state.moves.length - 1);

export const addMove = (state, board, fromSlot, toSlot, pending) => {
  const move = newMove({
    fromPos: logic.getRankFile(fromSlot),
    toPos: logic.getRankFile(toSlot),
    piece: board.getPiece(fromSlot),
    board: board.move(fromSlot, toSlot),
    pending,
  });
  return selectLastMove(
    incrementMoveCount(
      update(state, { moves: { $push: [move] } }),
    ),
  );
};

export const cancelMoves = (state) => {
  const { moves, selectedMoveId } = state;
  const confirmedMoves = moves.filter(({ pending }) => !pending);
  const selectedMoveIndex = Math.min(selectedMoveId, confirmedMoves.length - 1);

  return {
    ...state,
    moves: confirmedMoves,
    moveCount: confirmedMoves.length,
    selectedMoveId: setSelectedMoveIndex(state, selectedMoveIndex),
  };
};

export const confirmMoves = (state) => {
  const { moves } = state;
  return {
    ...state,
    moves: moves.map((move) => ({ ...move, pending: false })),
  };
};

const setMove = (state, { origin: fromPos, destination: toPos }) => {
  const { board } = state.moves[state.moves.length - 1];
  const move = newMove({
    piece: board.getPiece(fromPos, logic.RefType.RANK_FILE),
    fromPos,
    toPos,
    board: board.move(fromPos, toPos, logic.RefType.RANK_FILE),
    pending: false,
  });
  return incrementMoveCount(update(state, { moves: { $push: [move] } }));
};

export const setMoves = (state, moves) =>
  selectLastMove(
    moves.reduce(
      (prevState, move) => setMove(prevState, move),
      {
        ...state,
        moves: getInitialMovesState(),
        moveCount: 0,
        loading: false,
      },
    ),
  );

export default {};
