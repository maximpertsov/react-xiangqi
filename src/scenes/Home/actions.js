import * as client from 'services/client';

export const fetchGames = ({ username }) => async dispatch => {
  if (username === null) return;

  const {
    data: { games },
  } = await client.getGameList({ username });
  dispatch({ type: 'set_games', games });
};

export const setForm = form => dispatch => {
  const { username, password, error } = form;
  if (username !== undefined) {
    dispatch({ type: 'set_login_username', username });
  }
  if (password !== undefined) {
    dispatch({ type: 'set_login_password', password });
  }
  if (error !== undefined) {
    dispatch({ type: 'set_login_error', error });
  }
};

export default {};
