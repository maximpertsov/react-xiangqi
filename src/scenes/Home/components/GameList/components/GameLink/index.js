import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { Button } from 'semantic-ui-react';

const GameLink = ({ slug, setGameSlug }) => {
  const setThisGameSlug = useCallback(
    () => setGameSlug(slug),
    [slug, setGameSlug],
  );

  return (
    <Button
      onClick={setThisGameSlug}
      className="GameLink"
    >
      {slug}
    </Button>
  );
};

GameLink.propTypes = {
  slug: PropTypes.string.isRequired,
  setGameSlug: PropTypes.func.isRequired,
};

export default GameLink;
