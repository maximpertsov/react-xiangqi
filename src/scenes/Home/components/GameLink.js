import React from 'react';
import PropTypes from 'prop-types';
import { Button, Header } from 'semantic-ui-react';
import { useDispatch, useSelector } from 'react-redux';
import { createSelector } from 'reselect';

import isEqual from 'lodash/isEqual';
import get from 'lodash/get';

import BoardView from 'components/Board/components/BoardView';
import actions from 'actions';
import { getOpponent } from 'reducers/selectors';

const mapStateToProps = createSelector(
  state => state.username,
  (_, props) => props.game,

  (username, game) => ({
    opponent: getOpponent({ username, ...game }),
    teamBlackPOV: username && username === get(game, 'player2.name'),
  }),
);

const GameLink = ({ game }) => {
  const dispatch = useDispatch();

  const { opponent, teamBlackPOV } = useSelector(
    state => mapStateToProps(state, { game }),
    isEqual,
  );

  const setThisGameSlug = () => {
    dispatch(actions.game.slug.set(game.slug));
    dispatch(actions.home.showGame.set(true));
  };

  return (
    <Button onClick={setThisGameSlug} className="GameLink">
      <Header size="tiny">vs {opponent.name}</Header>
      <BoardView
        teamBlackPOV={teamBlackPOV}
        move={game.currentMove}
        size="tiny"
      />
    </Button>
  );
};

GameLink.propTypes = {
  game: PropTypes.shape().isRequired,
};

export default GameLink;
