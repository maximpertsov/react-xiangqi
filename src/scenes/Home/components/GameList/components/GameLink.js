import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { Button } from 'semantic-ui-react';
import { useDispatch } from 'react-redux';
import actions, { toggleShowGame } from 'actions';

import styled from '@emotion/styled';

const Wrapper = styled.div`
  padding: 5px;
`;

const GameLink = ({ slug }) => {
  const dispatch = useDispatch();

  const setThisGameSlug = useCallback(() => {
    dispatch(actions.game.slug.set(slug));
    dispatch(toggleShowGame({ showGame: true }));
  }, [dispatch, slug]);

  return (
    <Wrapper>
      <Button fluid onClick={setThisGameSlug} className="GameLink">
        {slug}
      </Button>
    </Wrapper>
  );
};

GameLink.propTypes = {
  slug: PropTypes.string.isRequired,
};

export default GameLink;
