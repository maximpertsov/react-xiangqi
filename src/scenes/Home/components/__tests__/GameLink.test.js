import React from 'react';
import { useDispatch } from 'react-redux';
import actions from 'actions';

import GameLink from '../GameLink';

jest.mock('react-redux');

describe('GameLink', () => {
  const slug = 'abc123';
  const store = mockStore({});

  let wrapper;

  beforeEach(() => {
    useDispatch.mockReturnValue(store.dispatch);
    wrapper = shallowWrappedComponent(<GameLink slug={slug} />, store);
  });

  afterEach(() => {
    store.clearActions();
    jest.resetAllMocks();
  });

  test('snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('click', () => {
    wrapper.find('Button').simulate('click');

    expect(store.getActions()).toStrictEqual([
      actions.game.slug.set(slug),
      actions.home.showGame.set(true),
    ]);
  });
});
