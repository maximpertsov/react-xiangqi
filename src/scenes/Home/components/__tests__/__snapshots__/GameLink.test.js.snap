// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameLink snapshot 1`] = `
<Button
  as="button"
  className="GameLink"
  onClick={[Function]}
>
  abc123
</Button>
`;
